timemachine/fe/single_topology_v3.py

```diff
@@ -177,10 +177,59 @@ def setup_dummy_interactions(
     )
 
 
+def canonicalize_improper_idxs(idxs):
+    """
+    Canonicalize an improper_idx while being symmetry aware.
+
+    Given idxs (i,j,k,l), where i is the center, and (j,k,l) are neighbors:
+
+    0) Canonicalize the (j,k,l) into (jj,kk,ll) by sorting
+    1) Generate clockwise rotations of (j,k,l)
+    2) Generate counter clockwise rotations of (j,k,l)
+    3) We now can sort 1) and 2) and assign a mapping
+
+    If the (j,k,l) is in the cw rotation ordered set, we're done. Otherwise it must
+    be in the ccw ordered set. We look up the corresponding idx in the cw set.
+
+    This does not do idxs[0] < idxs[-1] canonicalization.
+    """
+    i, j, k, l = idxs
+
+    # i is the center
+    # generate lexical order
+    key = (j, k, l)
+
+    jj, kk, ll = sorted(key)
+
+    # generate clockwise permutations
+    cw_jkl = (jj, kk, ll)
+    cw_klj = (kk, ll, jj)
+    cw_ljk = (ll, jj, kk)
+    cw_items = sorted([cw_jkl, cw_klj, cw_ljk])
+
+    if key in cw_items:
+        return (i, j, k, l)
+
+    # generate counter clockwise permutations
+    ccw_kjl = (kk, jj, ll)
+    ccw_jlk = (jj, ll, kk)
+    ccw_lkj = (ll, kk, jj)
+    ccw_items = sorted([ccw_kjl, ccw_jlk, ccw_lkj])
```

Comment:
Might be slightly cleaner to refactor this using two operations `swap(i, j, k) -> (j, i, k)`, `rotate(i, j, k) -> (j, k, i)`, rather than writing out all the permuted idxs?

```suggestion
    def rotate(i, j, k):
        return (j, k, i)

    def generate_rotations(i, j, k):
        orig = (i, j, k)
        rot1 = rotate(*orig)
        rot2 = rotate(*rot1)
        rotations = [orig, rot1, rot2]
        return sorted(rotations)

    cw_items = generate_rotations(i, j, k)

    if key in cw_items:
        return (i, j, k, l)

    ccw_items = generate_rotations(j, i, k)
```

(on the fence about this suggestion)

Reply:
For a simple case like this I prefer explicitly listing out the elements of the cw/ccw idxs, since the suggested version requires me to actually mentally work out what the elements are (even though its more semantically precise). I'll add a comment detailling the steps though.